### ISSUE TITLE
Pretraining: prepare a reproducible Aven baseline

### ISSUE THREAD

**Issue by ariveinberg-max** [ariveinberg-max/aven-1#1]
Owner: Claude (coordination label, not a GitHub account assignment).

Continue the existing Aven pretraining work. Record the dataset version/hash, tokenizer, architecture, seed, training command, W&B run URL and checkpoint location. Preserve existing checkpoints.

Completion criteria:
- [ ] Document the active run and its configuration.
- [ ] Save a baseline checkpoint with a reproducible evaluation command.
- [ ] Record training and held-out loss, token counts and compute used.
- [ ] Hand off a frozen checkpoint and tokenizer for evaluation.

Coordinate through research/TASKS.md. Do not run competing GPU jobs on the 8 GB Mac. This issue records work already reported by the user; it does not automatically notify or start Claude Code.

**Comment by ariveinberg-max** [comment]
## Pretraining baseline — status as of 2026-09-08

**Owner / issue:** Claude / #1

**Branch and files changed:** none (documentation only, on `main` at `8ad1d55`)

### Configuration
- Architecture: `Config(width=768, layers=8, heads=12, context=192, vocab=2048)` — 58,424,832 parameters
- Tokenizer: from-scratch BPE, vocab 2048, trained on the corpus below (3.474 bytes/token, 640,821 tokens)
- Optimizer/seed: AdamW, `torch.manual_seed(42)`, validation split uses `torch.Generator().manual_seed(123)`
- Dataset: `data/training.txt`, 2,226,235 bytes
  - SHA-256: `096ba20ef95ba56dbdddb87381dc86cc8328c179893fcbab5ee515aad5f68fac`
- Training command:
  ```
  .venv/bin/python train.py --data data/training.txt --steps 1000 --width 768 --layers 8 --heads 12 --vocab 2048 --context 192 --wandb
  ```
- Device: MPS (local M2, 8 GB)
- Run ID: `pretrain-096ba20e-52bb62a7`
- W&B run: https://wandb.ai/ariveinberg-ari-research/aven-1/runs/pretrain-096ba20e-52bb62a7

### Result (incomplete run, not the frozen baseline yet)
- Target: 1000 steps. Actual: killed by macOS at step 574/1000 due to system-wide memory pressure (multiple unrelated local apps competing for RAM on the 8 GB Mac) — not a training instability, no NaN/divergence.
- Loss trajectory (train / held-out / perplexity), from `checkpoints/status.json`:
  - step 0: held-out 7.747, perplexity 2314.77
  - step 200: held-out 6.828, perplexity 923.5
  - step 400: held-out 6.075, perplexity 434.94
  - step 560 (last full log line): train 5.753, held-out 5.924, perplexity 373.86
  - step 574 (last autosave, per status.json): train_loss 5.677
- Checkpoint auto-saved cleanly (checkpoints save every 20 steps); no steps lost beyond the last periodic save.

### Checkpoint and tokenizer location/hash
- `checkpoints/latest.pt` — SHA-256: `49e021f7d9f063af5555f35c15c79ff009a5b30444a2266e78d4a9d99b4823c3`
- `checkpoints/tokenizer.json` — SHA-256: `5dc2df20655b1221db002afb8a5f6fa5bbdcea660a2674cea11ab3cef9dbb91b`
- Both are local-only (gitignored), not yet uploaded anywhere durable/shared.

### Known limitations
- **Not the frozen baseline**: run is at step 574/1000, not the completion target in this issue. Held-out perplexity is still falling steeply (373.86 and dropping), so this checkpoint under-represents the intended baseline's quality.
- Held-out set is a held-out slice of the same small, homogeneous corpus, not a separate distribution — perplexity here should not be read as general capability.
- Resume attempt on Google Colab (to finish the remaining ~440 steps on free GPU, since local RAM is the bottleneck) is in progress and currently blocked: an earlier Colab attempt failed from a stale `brain.py` in the notebook (missing the `attn_scale` config field added after the notebook was first generated) and a stale/mismatched `data/training.txt` upload to Drive. Notebook was regenerated (`work/build_colab_notebook.py`) from current source to fix the first issue; corpus re-upload and Colab reconnect are still pending on the user's side as of this comment.

### Next action and owner
- Owner: Claude / user. Re-upload the correct `data/training.txt` (hash above) and the regenerated `Aven-1-Colab.ipynb` to Drive/Colab, reconnect a fresh runtime, and run `train.py --resume --steps 440 --device cuda --wandb` to reach the step-1000 target.
- Once that completes, this issue's checkpoint/tokenizer/hash values above should be replaced with the step-1000 frozen values before handoff to evaluation (#2).

**Comment by ariveinberg-max** [comment]
## Pretraining baseline — FROZEN (58M model)

Resumed on Google Colab (T4 GPU) from the local step-574 checkpoint after fixing two Colab-side issues: a stale notebook missing the `attn_scale` Config field, and a stale/mismatched `data/training.txt` on Drive. Both fixed by regenerating the notebook from current source (`work/build_colab_notebook.py`) and re-uploading the correct corpus file.

### Final state
- Step: **2100** (well past the original 1000-step target — `--steps N` on resume is additive to the current step, not absolute, so it overshot; stopped manually via Colab interrupt)
- Train loss: 4.510
- Held-out loss: 5.148
- Held-out perplexity: **172.07** (down from 2314.77 at step 0, 373.86 at step 560)
- Device for this stretch: CUDA (T4), vs. MPS for steps 0–574
- W&B run (same run ID throughout, local MPS + Colab CUDA merged into one history): https://wandb.ai/ariveinberg-ari-research/aven-1/runs/pretrain-096ba20e-52bb62a7

### Frozen artifacts
- `checkpoints/latest.pt` — SHA-256: `f510faba68ef14a30822c16c7394ca978b63b80a2c99d6221a9d3a304a427112`
- `checkpoints/tokenizer.json` — SHA-256: `5dc2df20655b1221db002afb8a5f6fa5bbdcea660a2674cea11ab3cef9dbb91b` (unchanged from earlier — same tokenizer, same corpus)
- Dataset: `data/training.txt`, SHA-256 `096ba20ef95ba56dbdddb87381dc86cc8328c179893fcbab5ee515aad5f68fac` (unchanged)
- Architecture: `Config(width=768, layers=8, heads=12, context=192, vocab=2048)`, 58,424,832 parameters
- Seed: `torch.manual_seed(42)`, validation split `Generator().manual_seed(123)`

### Known limitations (unchanged from prior comment)
- Held-out set is a held-out slice of the same small, homogeneous corpus — perplexity here is not a measure of general capability.
- This is a from-scratch byte/BPE causal LM at 58M params; no instruction-tuning or RLHF applied at this checkpoint (that pipeline exists separately on the earlier, smaller architecture and needs to be redone on this one if/when this becomes the working model).

### Next action and owner
- **This checkpoint is now frozen and ready for #2 (independent evaluation, Codex).**
- Claude is now starting a separate, larger 150M-parameter pretraining run (`--width 1024 --layers 12 --heads 16 --context 192`) on the same Colab GPU, as a distinct experiment — it does not touch or replace this frozen 58M checkpoint/hash above.

**Comment by ariveinberg-max** [comment]
## Kaggle blocker resolved (handoff from Codex, finished by Claude)

**Owner / issue:** Claude / #1

Codex's fix in 705e9e0 embedded `tracking.py` directly into both cloud notebooks (`work/build_colab_notebook.py`, `work/build_kaggle_notebook.py`) so they no longer need a separate import — but `tracking.py` itself and the `train.py` changes that use it were never committed, so a fresh clone would still fail to regenerate the notebooks.

**Branch and files changed:** `main` at `136211d` — committed `tracking.py`, `test_tracking.py`, `WANDB.md`, and the `train.py`/`README.md`/`runs/ledger.jsonl` changes that were sitting uncommitted.

**Verified:** `python -m unittest test_tracking` passes. Cloned the repo fresh into a scratch dir, ran both `work/build_*_notebook.py` generators with no errors, and diffed the output against the committed `.ipynb` files — byte-identical. The `ModuleNotFoundError` blocker is resolved and the repo is self-consistent from a clean checkout.

**Not done (needs the user's Kaggle login, can't be run from an agent session):** `checkpoints/status.json` still shows `step: 2100, target: 3560, status: "paused"` — the extended run to step 4000 (matched-compute comparison vs. the frozen 58M baseline) has not actually progressed on Kaggle yet.

**Next action and owner:** User — upload `Aven-1-Kaggle.ipynb` (now fixed) plus the `data/`/`checkpoints/` Kaggle Dataset per the notebook's own instructions, attach a GPU runtime, and run cell 6 (`--resume --steps <N> --device cuda --wandb`) to continue toward step 3560/4000.

**Comment by ariveinberg-max** [comment]
**Kaggle continuation incident record (2026-09-09; metrics only).** Ari reported that the 58,424,832-parameter continuation reached step **3560**, with train loss approximately **3.45** and held-out loss approximately **4.914**, before the checkpoint was lost with Kaggle's no-persistence session. No usable checkpoint from that endpoint has been verified; these numbers must not be presented as a resumable or independently reevaluated model.

A read-only recovery of the [shared W&B run](https://wandb.ai/ariveinberg-ari-research/aven-1/runs/pretrain-096ba20e-52bb62a7) found two distinct records: the history row at model step 3560 contains train loss **3.447188377380371**, held-out loss **4.912829279899597**, and held-out perplexity **136.02**; the final summary contains `final_step=3560` and `final_val_loss=4.914183259010315`. The summary and history are not a single consistent evaluation, and concurrent writers prevent unambiguous platform attribution from run ID alone. Preserve both; do not attach the history perplexity to the different final-summary loss. The old Kaggle notebook also retains a later assertion failure, `Expected step 2100; found 4000`; this is evidence of a reported checkpoint step at that time, not evidence of a surviving step-4000 artifact or its metrics. Step 3560 is therefore a preserved observation, not a proven maximum step. See [issue #1](https://github.com/ariveinberg-max/aven-1/issues/1). Equal optimizer steps across 58M and 153M models are a matched-step comparison, not matched compute.

Documentation-only update: preserving surviving observations without claiming a usable checkpoint. The earlier local status snapshot was not a complete account of remote Kaggle progress. Concurrent-resume coordination guidance has also been added to research/TASKS.md.
